Mathlib/NumberTheory/ModularForms/Basic.lean

```diff
@@ -397,3 +400,109 @@ instance (priority := 99) [CuspFormClass F Γ k] : ModularFormClass F Γ k where
   bdd_at_infty _ _ := (CuspFormClass.zero_at_infty _ _).boundedAtFilter
 
 end CuspForm
+
+namespace ModularForm
+
+section GradedRing
+
+/--cast for modular forms, which is useful for removing `heq`'s.-/
+def mcast {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (h : a = b) (f : ModularForm Γ a) : ModularForm Γ b
+    where
+  toFun := (f : ℍ → ℂ)
+  slash_action_eq' := by intro A; have := f.slash_action_eq' A; convert this; exact h.symm
+  holo' := f.holo'
+  bdd_at_infty' := by intro A; convert f.bdd_at_infty' A <;> exact h.symm
+
+@[ext]
+theorem gradedMonoid_eq_of_cast {Γ : Subgroup SL(2, ℤ)} {a b : GradedMonoid (ModularForm Γ)}
+    (h : a.fst = b.fst) (h2 : mcast h a.snd = b.snd) : a = b := by
+  obtain ⟨i, a⟩ := a
+  obtain ⟨j, b⟩ := b
+  cases h
+  exact congr_arg _ h2
+
+theorem one_mul (k : ℤ) {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ k) :
+    mcast (zero_add k) ((1 : ModularForm Γ 0).mul f) = f := by
+  rw [mcast]
+  ext1
+  simp
+  rfl
+
+theorem mul_one (k : ℤ) {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ k) :
+    mcast (add_zero k) (f.mul (1 : ModularForm Γ 0)) = f := by
+  rw [mcast]
+  ext1
+  simp
+  rfl
+
+theorem mul_assoc {a b c : ℤ} {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ a)
+    (g : ModularForm Γ b) (h : ModularForm Γ c) :
+  mcast (add_assoc _ _ _) ((f.mul g).mul h) = (f.mul (g.mul h)) := by
+  rw [mcast]
+  congr
+  ext1
+  simp only [mul_coe, Pi.mul_apply, toSlashInvariantForm_coe]
+  ring
+
+theorem mul_comm {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ a) (g : ModularForm Γ b) :
+    mcast (add_comm a b) (f.mul g) = (g.mul f) := by
+  rw [mcast]
+  congr
+  ext1
+  simp only [mul_coe, Pi.mul_apply, toSlashInvariantForm_coe]
+  ring
+
+instance (Γ : Subgroup SL(2, ℤ)) : GradedMonoid.GOne (ModularForm Γ) where
+  one := 1
+
+instance (Γ : Subgroup SL(2, ℤ)) : GradedMonoid.GMul (ModularForm Γ) where
+  mul f g := f.mul g
+
+open GradedMonoid
+
+instance (Γ : Subgroup SL(2, ℤ)) : NatCast (ModularForm Γ 0) where
+  natCast := fun n => n • (1 : ModularForm Γ 0)
+
+instance (Γ : Subgroup SL(2, ℤ)) : IntCast (ModularForm Γ 0) where
+  intCast := fun n => n • (1 : ModularForm Γ 0)
+
+lemma MF_intcast_eq_SIF_intcast  (Γ : Subgroup SL(2, ℤ)) (n : ℤ) :
+    (n : ModularForm Γ 0) =  (n :  SlashInvariantForm Γ 0) := by
+  rfl
+
+@[simp]
+lemma natCast_coe (Γ : Subgroup SL(2, ℤ)) (n : ℕ) :
+    (n : ModularForm Γ 0 ) = n • (1 : ModularForm Γ 0) := by rfl
+
+@[simp]
+lemma intCast_coe (Γ : Subgroup SL(2, ℤ)) (n : ℤ) :
+    (n : ModularForm Γ 0 ) = n • (1 : ModularForm Γ 0) := by rfl
```

Comment:
Ideally the statement of these would be
```suggestion
@[simp]
lemma natCast_coe (Γ : Subgroup SL(2, ℤ)) (n : ℕ) :
    ⇑(n : ModularForm Γ 0 ) = n := rfl

@[simp]
lemma intCast_coe (Γ : Subgroup SL(2, ℤ)) (z : ℤ) :
    ⇑(z : ModularForm Γ 0) = z := rfl
```
Note that `by rfl` is almost always an antipattern, you should prefer plain `rfl` unless it times out.